[lldb-dap] Refactoring lldb-dap 'launch' request to use typed RequestHandler<>.

ashgti opened the pull request:
This converts a number of json::Value's into well defined types that are used throughout lldb-dap and updates the 'launch' command to use the new well defined types.

ashgti committed:
[lldb-dap] Refactoring lldb-dap 'launch' request to use typed RequestHandler<>.

This converts a number of json::Value's into well defined types that are used throughout lldb-dap and updates the 'launch' command to use the new well defined types.

ashgti commented:
I may need to split this up into a smaller set of patches. Let me know what you think.

ashgti committed:
Merge branch 'main' into launch-req

ashgti commented:
Split the DAPConfiguration structure into its own patch #133960 to make this a bit smaller.

ashgti committed:
Merge remote-tracking branch 'origin/main' into launch-req

ashgti committed:
Fixing up after merging.

ashgti committed:
Merge branch 'launch-req' of github.com:ashgti/llvm-project into launch-req

ashgti committed:
Apply clang-format.

ashgti committed:
Apply python format.

JDevlieghere commented on lldb/tools/lldb-dap/Handler/LaunchRequestHandler.cpp:
```diff
@@ -9,127 +9,66 @@
+  if (dap.target.GetProcess().IsValid()) {
     if (dap.is_attach)
-      SendProcessEvent(dap, Attach); // this happens when doing runInTerminal
+      // this happens when doing runInTerminal
+      SendProcessEvent(dap, Attach);
```
With the comment moved, this (and the else-case) now need braces 🙃

ashgti replied:
Adjusted this to use a ternary if and moved the comment.

ashgti committed:
Moving the defaulted values to optional.

ashgti committed:
Adjusting a comment block and associated if block.

ashgti commented on lldb/tools/lldb-dap/Handler/RequestHandler.cpp:
```diff
@@ -33,15 +35,12 @@ MakeArgv(const llvm::ArrayRef<std::string> &strs) {
+  else
+    flags &= ~mask;
   return flags;
 }
```
Converted the bools back to optionals.

ashgti replied:
I'm not sure I understand this issue. For example, the `disableASLR`. If a user launched with:

```json
{
  "type": "lldb-dap",
  "request": "launch",
  "name": "Debug a.out",
  "program": "a.out"
}
```

This would default to `stopOnEntry == false`.

If the user launched with:

```json
{
  "type": "lldb-dap",
  "request": "launch",
  "name": "Debug a.out",
  "program": "a.out",
  "stopOnEntry": false
}
```

This would also set `stopOnEntry == false`.

If a user launched with:

```json
{
  "type": "lldb-dap",
  "request": "launch",
  "name": "Debug a.out",
  "program": "a.out",
  "stopOnEntry": true
}
```

This would result in `stopOnEntry == true`.

I guess the only complicated example is the `disableASLR` flag because that is enabled by default because of https://github.com/llvm/llvm-project/blob/af63e1b505453de3e6a281d1b72e62fa8d396b23/lldb/source/API/SBLaunchInfo.cpp#L47 so its flag that is defaulted to `true` implicitly.

ashgti committed:
Merge branch 'main' into launch-req

ashgti committed:
Applying python format.

ashgti committed:
Apply clang-format to organize imports.

ashgti committed:
Merge branch 'main' into launch-req

ashgti committed:
Moving 'program', 'targetTriple' and 'platform' to the protocol::Configuration since they're shared between launch and attach.

Additionally, I tried to be more consistent with how we apply default flags state in the protocol definition.

ashgti committed:
Merge branch 'main' into launch-req

ashgti committed:
Fixing "launch" request handler post merge.

ashgti commented:
I updated my branch with main and updated the launch handler to return an `llvm::Error` to simplify the body.

ashgti committed:
Apply suggestions from code review

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

ashgti committed:
Adjusting the error handling for launches and adding extra tests for additional launch failure cases.

ashgti commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -672,21 +670,17 @@ DAP::CreateTargetFromArguments(const llvm::json::Object &arguments,
     error.SetErrorStringWithFormat(
-        "Could not create a target for a program '%s': %s.", program.data(),
-        error.GetCString());
+        "Could not create a target for a program: %s.", error.GetCString());
```
Previously the error message would print something along the lines of:

```
Could not create a target for a program 'a.out': 'a.out' does not exist.
```

I'm not sure the 'Could not create target' part adds anything, removing it the error message becomes:

```
'a.out' does not exist
```

Here is an example using:

```
        {
            "type": "lldb-dap",
            "request": "launch",
            "name": "Debug bad path",
            "program": "~/Projects/baz-bad-path"
        },
```

<img width="265" alt="Screenshot 2025-04-24 at 5 25 44 PM" src="https://github.com/user-attachments/assets/a37c5052-0d14-4a5d-8e8c-92d5ba974ca3" />

JDevlieghere replied:
Agreed

ashgti commented on lldb/tools/lldb-dap/Handler/LaunchRequestHandler.cpp:
```diff
@@ -9,127 +9,69 @@
+  const std::string debuggerRoot = dap.configuration.debuggerRoot.value_or("");
   if (!debuggerRoot.empty())
-    llvm::sys::fs::set_current_path(debuggerRoot);
+    sys::fs::set_current_path(debuggerRoot);
```
With bazel/blaze remote builds, when we build things for Apple platforms we use `-oso_prefix .` for the debugger paths. So setting this correctly is important for the relative paths to resolve correctly. I think there is a new `target.object-map` setting in lldb, but I don't think that exists in any version of lldb shipped with Xcode yet.

ashgti commented on lldb/tools/lldb-dap/Handler/LaunchRequestHandler.cpp:
```diff
@@ -9,127 +9,69 @@
-  }
+  dap.SetTarget(dap.CreateTarget(status));
+  if (status.Fail())
+    return make_error<DAPError>(status.GetCString());
```
https://github.com/llvm/llvm-project/blob/main/lldb/tools/lldb-dap/LLDBUtils.h#L159-L160 We have a helper for that today, updates to use that.

ashgti commented on lldb/tools/lldb-dap/Handler/RequestHandler.cpp:
```diff
@@ -37,14 +39,12 @@ MakeArgv(const llvm::ArrayRef<std::string> &strs) {
-  }
+static uint32_t SetLaunchFlag(uint32_t flags, bool flag,
+                              lldb::LaunchFlags mask) {
+  if (flag)
+    flags |= mask;
+  else
+    flags &= ~mask;
```
I'm not sure I understand this issue. For example, the `disableASLR`. If a user launched with:

```json
{
  "type": "lldb-dap",
  "request": "launch",
  "name": "Debug a.out",
  "program": "a.out"
}
```

This would default to `stopOnEntry == false`.

If the user launched with:

```json
{
  "type": "lldb-dap",
  "request": "launch",
  "name": "Debug a.out",
  "program": "a.out",
  "stopOnEntry": false
}
```

This would also set `stopOnEntry == false`.

If a user launched with:

```json
{
  "type": "lldb-dap",
  "request": "launch",
  "name": "Debug a.out",
  "program": "a.out",
  "stopOnEntry": true
}
```

This would result in `stopOnEntry == true`.

I guess the only complicated example is the `disableASLR` flag because that is enabled by default because of https://github.com/llvm/llvm-project/blob/af63e1b505453de3e6a281d1b72e62fa8d396b23/lldb/source/API/SBLaunchInfo.cpp#L47 so its flag that is defaulted to `true` implicitly. But I made the default value `true` for that field, so it should be handled correctly.

JDevlieghere replied:
Right, so in the case where the user didn't specify anything, I'm arguing we shouldn't treat it the same as if it was set to false. Unless the spec says that it being absent should be treated as it being set to false. The code snippet you're pointing out is exactly what I'm worried about: some of these things are set by default and the implementation here shouldn't have to worry about that.

ashgti replied:
For the launch.json, we're the ones defining the fields and their defaults, defined in the package.json https://github.com/llvm/llvm-project/blob/5dc2d668e68613e8898c5c42960d792129453e9d/lldb/tools/lldb-dap/package.json#L226-L250 

The defaults I'm specifying in the `LaunchRequestArguments` are setup to match those defaults as well. We can change the defaults, since this part is defined by the lldb-dap itself.

JDevlieghere replied:
That's the part I was missing, thanks for clarifying. I can't say I'm particularly happy about having to manually keep things in sync, but it does remove the need for the additional complexity that the optionals introduce.

ashgti replied:
Technically, whatever lldb-dap does is the source of through here. I think the package.json description is actually out of date for a few fields. I'll send a PR to fix those.

ashgti committed:
Applying formatting.

DavidSpickett commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -979,8 +967,8 @@ lldb::SBError DAP::WaitForProcessToStop(uint32_t seconds) {
   }
-  error.SetErrorStringWithFormat("process failed to stop within %u seconds",
-                                 seconds);
+  error.SetErrorStringWithFormat("process failed to stop within %lld seconds",
```
Warning on AArch64 Linux compiling with clang:
```
[665/779] Building CXX object tools/lldb/tools/lldb-dap/CMakeFiles/lldb-dap.dir/DAP.cpp.o
/home/david.spickett/llvm-project/lldb/tools/lldb-dap/DAP.cpp:972:34: warning: format specifies type 'long long' but the argument has type 'rep' (aka 'long') [-Wformat]
  971 |   error.SetErrorStringWithFormat("process failed to stop within %lld seconds",
      |                                                                 ~~~~
      |                                                                 %ld
  972 |                                  seconds.count());
      |                                  ^~~~~~~~~~~~~~~
1 warning generated.
```
